pkg/appregistry/downloader_test.go

```diff
@@ -0,0 +1,132 @@
+package appregistry
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/operator-framework/operator-registry/pkg/apprclient"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+var testPrepare = []struct {
+	input                 *Input
+	sourceQuerier         sourceQuerier
+	expectedDownloadItems []*downloadItem
+	expectedError         utilerrors.Aggregate
+}{
+	{
+		input: &Input{
+			Sources: []*Source{
+				{
+					Endpoint:          "quay.io",
+					RegistryNamespace: "",
+				},
+			},
+			Packages: []string{
+				"Kubevirt",
+				"etcd",
+			},
+		},
+		sourceQuerier: &fakeSourceQuerier{
+			map[Source][]*apprclient.RegistryMetadata{
+				Source{
+					Endpoint:          "quay.io",
+					RegistryNamespace: "",
+				}: {
+					&apprclient.RegistryMetadata{Name: "Kubevirt"},
+				},
+			},
+			map[Source]error{},
+		},
+		expectedDownloadItems: []*downloadItem{
+			&downloadItem{
+				&apprclient.RegistryMetadata{Name: "Kubevirt"},
+				&Source{Endpoint: "quay.io", RegistryNamespace: ""},
+			},
+		},
+		expectedError: nil,
+	},
+	{
+		input: &Input{
+			Sources: []*Source{
+				{
+					Endpoint:          "quay.io",
+					RegistryNamespace: "",
+				},
+				{
+					Endpoint:          "other-endpoint.io",
+					RegistryNamespace: "",
+				},
+			},
+			Packages: []string{
+				"Kubevirt",
+				"etcd",
+			},
+		},
+		sourceQuerier: &fakeSourceQuerier{
+			map[Source][]*apprclient.RegistryMetadata{
+				Source{
+					Endpoint:          "quay.io",
+					RegistryNamespace: "",
+				}: {
+					&apprclient.RegistryMetadata{Name: "Kubevirt"},
+				},
+			},
+			map[Source]error{
+				Source{
+					Endpoint:          "other-endpoint.io",
+					RegistryNamespace: "",
+				}: errors.New("Failed to fetch sources from other-endpoint.io"),
+			},
+		},
+		expectedDownloadItems: []*downloadItem{
+			&downloadItem{
+				&apprclient.RegistryMetadata{Name: "Kubevirt"},
+				&Source{Endpoint: "quay.io", RegistryNamespace: ""},
+			},
+		},
+		expectedError: utilerrors.NewAggregate(
+			[]error{
+				errors.New("Failed to fetch sources from other-endpoint.io"),
+			},
+		),
+	},
+}
+
+type fakeSourceQuerier struct {
```

Comment:
In this case no, since I don't think that counting the interactions with the mock will make testing better. I prefer tests that assert the return value for a given input more than tests that count calls to mock objects.